website/addons/base/serializer.py

```diff
@@ -0,0 +1,210 @@
+            for each in self.user_settings.external_accounts
+        ]
+
+    @property
+    def serialized_user_settings(self):
+        retval = super(OAuthAddonSerializer, self).serialized_user_settings
+
+        retval['accounts'] = self.serialized_accounts
+
+        return retval
+
+    def serialize_account(self, external_account):
+        if external_account is None:
+            return None
+        return {
+            'id': external_account._id,
+            'provider_id': external_account.provider_id,
+            'display_name': external_account.display_name,
+            'nodes': [
+                self.serialize_granted_node(node)
+                for node in self.user_settings.get_attached_nodes(
+                    external_account=external_account
+                )
+            ]
+        }
+
+    @collect_auth
+    def serialize_granted_node(self, node, auth):
+
+        node_settings = node.get_addon(
+            self.user_settings.oauth_provider.short_name
+        )
+        serializer = node_settings.serializer(node_settings=node_settings)
+        urls = serializer.addon_serialized_urls
+        urls['view'] = node.url
+
+        return {
+            'id': node._id,
+            'title': node.title if node.can_view(auth) else None,
+            'urls': urls,
+        }
+
+class CitationsAddonSerializer(OAuthAddonSerializer):
+
+    REQUIRED_URLS = ['importAuth', 'folders', 'config', 'deauthorize', 'accounts']
+
+    '''
+    def _serialize_account(self, external_account):
+        if external_account is None:
+            return None
+        return {
+            'id': external_account._id,
+            'provider_id': external_account.provider_id,
+            'display_name': external_account.display_name,
+        }
+    '''
+
+    '''
+    @abc.abstractproperty
+    def serialized_model(self):
+        pass
+    '''
+
+    @property
+    def serialized_urls(self):
+        external_account = self.node_settings.external_account
+        ret = {
+            'auth': api_url_for('oauth_connect',
+                                service_name=self.node_settings.provider_name),
+            'settings': web_url_for('user_addons'),
+            'files': self.node_settings.owner.url,
+        }
+        if external_account and external_account.profile_url:
+            ret['owner'] = external_account.profile_url
+
+        addon_urls = self.addon_serialized_urls
+        # Make sure developer returns set of needed urls
+        for url in self.REQUIRED_URLS:
+            assert url in addon_urls, "addon_serilized_urls must include key '{0}'".format(url)
+        ret.update(addon_urls)
+        return ret
+
+    @property
+    def serialized_node_settings(self):
+        result = super(CitationsAddonSerializer, self).serialized_node_settings
+        result['folder'] = self.node_settings.selected_folder_name
+        return result
+
+    @property
+    def has_valid_credentials(self):
+        # TODO
```

Comment:
I would either like to see this implemented as an abc.abstractproperty, or removed altogether. Some addons exposed this as a way to pre-check that credentials are ready to be used with the provider's API (i.e. not expired or otherwise broken), and this gets used on the front end.

It may be an unnecessary check since we're already lazy-loading config, but this should be discussed.